File: sdk/python/tennetctl/vault.py
from __future__ import annotations

from typing import TYPE_CHECKING, Any

if TYPE_CHECKING:
    from ._transport import Transport


class VaultSecrets:
    """Vault secrets — get/list/create/rotate/delete.

    The SDK never caches plaintext. `get_secret` is a hot-path call; the
    backend's own 60s SWR cache in VaultClient handles the perf concern.
    """

    def __init__(self, transport: Transport) -> None:
        self._t = transport

    async def list(self, **filters: Any) -> list[dict]:
        params = {k: v for k, v in filters.items() if v is not None}
        data = await self._t.request("GET", "/v1/vault", params=params or None)
        return data if isinstance(data, list) else list(data or [])

    async def get(self, key: str) -> dict:
        """Retrieve a secret. Returns the envelope including the plaintext
        value — callers must handle the value carefully and never log it."""
        return await self._t.request("GET", f"/v1/vault/{key}")

    async def create(
        self,
        *,
        key: str,
        value: str,
        description: str | None = None,
    ) -> dict:
        body: dict[str, Any] = {"key": key, "value": value}
        if description is not None:
            body["description"] = description
        return await self._t.request("POST", "/v1/vault", json=body)

    async def rotate(self, key: str, *, value: str) -> dict:
        return await self._t.request(
            "POST", f"/v1/vault/{key}/rotate", json={"value": value}
        )

    async def delete(self, key: str) -> None:
        await self._t.request("DELETE", f"/v1/vault/{key}")


class VaultConfigs:
    """Vault configs — non-secret, operator-visible configuration keys."""

    def __init__(self, transport: Transport) -> None:
        self._t = transport

    async def list(self, **filters: Any) -> list[dict]:
        params = {k: v for k, v in filters.items() if v is not None}
        data = await self._t.request("GET", "/v1/vault-configs", params=params or None)
        return data if isinstance(data, list) else list(data or [])

    async def get(self, config_id: str) -> dict:
        return await self._t.request("GET", f"/v1/vault-configs/{config_id}")

    async def create(self, body: dict) -> dict:
        return await self._t.request("POST", "/v1/vault-configs", json=body)

    async def update(self, config_id: str, patch: dict) -> dict:
        return await self._t.request(
            "PATCH", f"/v1/vault-configs/{config_id}", json=patch
        )

    async def delete(self, config_id: str) -> None:
        await self._t.request("DELETE", f"/v1/vault-configs/{config_id}")


class Vault:
    def __init__(self, transport: Transport) -> None:
        self.secrets = VaultSecrets(transport)
        self.configs = VaultConfigs(transport)

File: sdk/typescript/src/vault.ts
import type { Transport } from "./transport.js";

export class VaultSecrets {
  constructor(private readonly t: Transport) {}

  async list(
    filters: Record<string, string | number | boolean | undefined> = {},
  ): Promise<Array<Record<string, unknown>>> {
    const params: Record<string, string | number | boolean | undefined> = {};
    for (const [k, v] of Object.entries(filters)) if (v !== undefined) params[k] = v;
    const data = await this.t.request<Array<Record<string, unknown>>>("GET", "/v1/vault", {
      params: Object.keys(params).length ? params : undefined,
    });
    return Array.isArray(data) ? data : [];
  }

  async get(key: string): Promise<Record<string, unknown>> {
    return this.t.request("GET", `/v1/vault/${key}`);
  }

  async create(args: {
    key: string;
    value: string;
    description?: string;
  }): Promise<Record<string, unknown>> {
    const body: Record<string, unknown> = { key: args.key, value: args.value };
    if (args.description !== undefined) body.description = args.description;
    return this.t.request("POST", "/v1/vault", { body });
  }

  async rotate(key: string, args: { value: string }): Promise<Record<string, unknown>> {
    return this.t.request("POST", `/v1/vault/${key}/rotate`, { body: { value: args.value } });
  }

  async delete(key: string): Promise<void> {
    await this.t.request("DELETE", `/v1/vault/${key}`);
  }
}

export class VaultConfigs {
  constructor(private readonly t: Transport) {}

  async list(
    filters: Record<string, string | number | boolean | undefined> = {},
  ): Promise<Array<Record<string, unknown>>> {
    const params: Record<string, string | number | boolean | undefined> = {};
    for (const [k, v] of Object.entries(filters)) if (v !== undefined) params[k] = v;
    const data = await this.t.request<Array<Record<string, unknown>>>("GET", "/v1/vault-configs", {
      params: Object.keys(params).length ? params : undefined,
    });
    return Array.isArray(data) ? data : [];
  }

  async get(configId: string): Promise<Record<string, unknown>> {
    return this.t.request("GET", `/v1/vault-configs/${configId}`);
  }

  async create(body: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.t.request("POST", "/v1/vault-configs", { body });
  }

  async update(configId: string, patch: Record<string, unknown>): Promise<Record<string, unknown>> {
    return this.t.request("PATCH", `/v1/vault-configs/${configId}`, { body: patch });
  }

  async delete(configId: string): Promise<void> {
    await this.t.request("DELETE", `/v1/vault-configs/${configId}`);
  }
}

export class Vault {
  readonly secrets: VaultSecrets;
  readonly configs: VaultConfigs;
  constructor(t: Transport) {
    this.secrets = new VaultSecrets(t);
    this.configs = new VaultConfigs(t);
  }
}
